src/mame/nintendo/nes_m82.cpp

```diff
@@ -0,0 +1,235 @@
+{
+	u8 ret = 0x40;
+
+	ret |= m_ctrl[0]->read_bit0();
+	ret |= m_ctrl[0]->read_bit34();
+	ret |= m_ctrl[1]->read_bit0();
+	ret |= m_ctrl[1]->read_bit34();
+
+	return ret;
+}
+
+u8 m82_state::m82_in1_r()
+{
+	u8 ret = 0x40;
+
+	ret |= m_ctrl[2]->read_bit0();
+	ret |= m_ctrl[2]->read_bit34();
+	ret |= m_ctrl[3]->read_bit0();
+	ret |= m_ctrl[3]->read_bit34();
+	ret |= m_ctrl[4]->read_bit0();
+	ret |= m_ctrl[4]->read_bit34();
+
+	return ret;
+}
+
+void m82_state::m82_in0_w(u8 data)
+{
+	for (int i = 0; i < 5; i++)
+		m_ctrl[i]->write(data);
+}
+
+
+/**************************************************************************/
+
+TIMER_CALLBACK_MEMBER(m82_state::m82_play_timer_cb)
+{
+}
+
+/***************************************************************************
+
+   Memory map
+
+***************************************************************************/
+
+void m82_state::m82_map(address_map &map)
+{
+	map(0x0000, 0x07ff).mirror(0x1800).ram();
+	map(0x2000, 0x3fff).rw(m_ppu, FUNC(ppu2c0x_device::read), FUNC(ppu2c0x_device::write));
+	map(0x4014, 0x4014).lw8(NAME([this](address_space &space, u8 data) { m_ppu->spriteram_dma(space, data); }));
+	map(0x4016, 0x4016).rw(FUNC(m82_state::m82_in0_r), FUNC(m82_state::m82_in0_w)); // IN0 - input port 1
+	map(0x4017, 0x4017).r(FUNC(m82_state::m82_in1_r));     // IN1 - input port 2 / PSG second control register
+
+	map(0x8000, 0xffff).rom(); // TODO: cart PRG muxed here
+}
+
+void m82_state::m82_ppu_map(address_map &map)
+{
+	map(0x0000, 0x1fff).rom(); // TODO: cart CHR muxed here
+	map(0x2000, 0x23ff).mirror(0x1000).bankrw(m_nt_page[0]);
+	map(0x2400, 0x27ff).mirror(0x1000).bankrw(m_nt_page[1]);
+	map(0x2800, 0x2bff).mirror(0x1000).bankrw(m_nt_page[2]);
+	map(0x2c00, 0x2fff).mirror(0x1000).bankrw(m_nt_page[3]);
+	map(0x3f00, 0x3fff).rw(m_ppu, FUNC(ppu2c0x_device::palette_read), FUNC(ppu2c0x_device::palette_write));
+}
+
+
+
+static INPUT_PORTS_START( nes_m82 )
+	PORT_START("CN13")
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_OTHER ) PORT_NAME("Game Select") PORT_CODE( KEYCODE_0 ) PORT_CHANGED_MEMBER(DEVICE_SELF, m82_state, m82_game_select, 0)
+
+	PORT_START("CN12")
+	PORT_CONFNAME( 0x0f, 0x08, "Play Time Limit" )
+	PORT_CONFSETTING(    0x01, ":30 sec." )
+	PORT_CONFSETTING(    0x02, "3:00 min." )
+	PORT_CONFSETTING(    0x04, "6:00 min." )
+	PORT_CONFSETTING(    0x08, "128:00 min." )
+INPUT_PORTS_END
+
+
+
+void m82_state::machine_start()
+{
+	m_nt_ram = std::make_unique<u8[]>(0x800);
+
+	for (int i = 0; i < 4; i++)
+		m_nt_page[i]->configure_entries(0, 2, m_nt_ram.get(), 0x400);
+
+	// vertical mirroring. TODO: replace this with proper code
+	m_nt_page[1]->set_entry(1);
+	m_nt_page[3]->set_entry(1);
+
+	m_play_timer = timer_alloc(FUNC(m82_state::m82_play_timer_cb), this);
+}
+
+void m82_state::machine_reset()
+{
+	switch (m_cn12->read())
+	{
+		case 1: m_time_limit =      20; break;
+		case 2: m_time_limit =  3 * 60; break;
+		case 4: m_time_limit =  6 * 60; break;
+		case 8: m_time_limit = 25 * 60; break;
```

Comment:
Is this supposed to match the DIP switch setting labels?  If it isn’t, you should probably explain why.

Reply:
Copy/paste error.